program-runtime/src/invoke_context.rs

```diff
@@ -522,75 +525,84 @@ impl<'a> InvokeContext<'a> {
     }
 
     /// Helper to prepare for process_instruction()
-    pub fn create_message(
+    #[allow(clippy::type_complexity)]
+    pub fn prepare_instruction(
         &mut self,
         instruction: &Instruction,
         signers: &[Pubkey],
-    ) -> Result<(Message, Vec<bool>, Vec<usize>), InstructionError> {
+    ) -> Result<(Vec<InstructionAccount>, Vec<bool>, Vec<usize>), InstructionError> {
+        // Normalize / unify the privileges of duplicate accounts by
+        // constructing a message and compiling an instruction
+        // only to throw them away immediately after.
         let message = Message::new(&[instruction.clone()], None);
-
-        // Gather keyed_accounts in the order of message.account_keys
-        let caller_keyed_accounts = self.get_instruction_keyed_accounts()?;
-        let callee_keyed_accounts = message
-            .account_keys
+        let instruction_accounts = message.instructions[0]
+            .accounts
```

Comment:
@Lichtso The consensus issue is that this used to iterate unique keys of the message, but now iterates all the keys in the instruction which could include duplicates.  Since it now iterates duplicates we may over charge the compute budget for each duplicate.  This would normally slip under the radar but for programs that are near the top of their compute budget, it could push them over the edge and cause them to fail.

Reply:
So it was directly related to the compute meter consumption in the runtime after all.
Thanks for figuring this out!